standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1561,7 +1561,7 @@ public void commitTxn(CommitTxnRequest rqst) throws NoSuchTxnException, TxnAbort
           assert true;
         }
 
-        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn) {
+        if (txnType != TxnType.READ_ONLY && !isReplayedReplTxn && txnType != TxnType.REBALANCE_COMPACTION) {
```

Comment:
why don't we do the same for TxnType.COMPACTION and instead have the if block `AND \"TC_OPERATION_TYPE\" <> " + OperationType.COMPACT` under moveTxnComponentsToCompleted?

Reply:
i think that is because we didn't have TxnType.COMPACTION type before the original change

Reply:
Normal compaction is not blocking (Uses SHARED_READ lock) so there is no need for conflict detection. Rebalance compaction uses EXCL_WRITE lock, and the records must not moved to the complete table in order to be able to detect conflicting write attempts. This is required to force concurrent updates/deletes to re-create their plan and use the new base/deltas containing the changed ACID rows.